File: solutions/c/kindergarten-garden/1/kindergarten_garden.c
#include "kindergarten_garden.h"
#include <string.h>

static plant_t char2plant(char c);

plants_t plants(const char *diagram, const char *student) {
    int st_idx = (student[0] - 'A') * 2;
    const char *row2 = strchr(diagram, '\n') + 1;
    
    plants_t pl;
    pl.plants[0] = char2plant(diagram[st_idx]);
    pl.plants[1] = char2plant(diagram[st_idx + 1]);
    pl.plants[2] = char2plant(row2[st_idx]);
    pl.plants[3] = char2plant(row2[st_idx + 1]);
    
    return pl;
}

static plant_t char2plant(char c) {
    switch (c) {
        case 'G':
            return GRASS;
        case 'C':
            return CLOVER;
        case 'R':
            return RADISHES;
        case 'V':
            return VIOLETS;
    }

    return -1;
}

File: solutions/csharp/kindergarten-garden/1/KindergartenGarden.cs
using System;
using System.Collections.Generic;

public enum Plant
{
    Violets,
    Radishes,
    Clover,
    Grass
}

public class KindergartenGarden
{
    string[] plants;
    
    public KindergartenGarden(string diagram)
    {
        plants = diagram.Split('\n');
    }

    public IEnumerable<Plant> Plants(string student)
    {
        var studentIndex = student[0] - 'A';

        var studentPlants = plants[0].Substring(studentIndex * 2, 2) + 
                            plants[1].Substring(studentIndex * 2, 2);

        foreach (var plant in studentPlants)
            yield return plant switch
            {
                'V' => Plant.Violets,
                'R' => Plant.Radishes,
                'C' => Plant.Clover,
                'G' => Plant.Grass,
                _ => throw new ArgumentException(),
            };
    }
}